src/Graphing/Debug.hs

```diff
@@ -0,0 +1,103 @@
+{-# LANGUAGE RecordWildCards #-}
+
+module Graphing.Debug (
+  GraphDebug (toDebug),
+  debugGraph,
+  renderGvFile,
+) where
+
+import Algebra.Graph.Acyclic.AdjacencyMap qualified as Acyclic
+import Algebra.Graph.AdjacencyMap qualified as Cyclic
+import Control.Effect.Lift (Has, Lift, sendIO)
+import Data.Bifunctor (Bifunctor (bimap))
+import Data.Set (Set)
+import Data.Set qualified as Set
+import Data.String (IsString)
+import Data.Text (Text)
+import Data.Text qualified as Text
+import Data.Text.Extra (showT)
+import Data.Text.IO qualified as TIO
+import DepTypes (DepEnvironment, Dependency (..), VerConstraint (..))
+import Graphing (Graphing (unGraphing), Node (..))
+
+gvFileName :: (IsString a) => a
+gvFileName = "./fossa.debug.gv"
+
+class Ord a => GraphDebug a where
+  toDebug :: a -> Text
+
+instance GraphDebug a => GraphDebug (Node a) where
+  toDebug Root = "ROOT"
+  toDebug (Node a) = toDebug a
+
+instance GraphDebug Dependency where
+  toDebug Dependency{..} = Text.unwords [dependencyName, version, envs]
+    where
+      version = maybe "*" versionTxt dependencyVersion
+      envs = "[" <> envsToText dependencyEnvironments <> "]"
+
+envsToText :: Set DepEnvironment -> Text
+envsToText = Text.intercalate ", " . map showT . Set.toList
+
+versionTxt :: VerConstraint -> Text
+versionTxt = \case
+  CEq txt -> txt
+  CURI txt -> txt
+  CCompatible txt -> txt
+  CAnd vc vc' -> versionTxt vc <> "AND" <> versionTxt vc'
+  COr vc vc' -> versionTxt vc <> "OR" <> versionTxt vc'
+  CLess txt -> txt
+  CLessOrEq txt -> txt
+  CGreater txt -> txt
+  CGreaterOrEq txt -> txt
+  CNot txt -> txt
+
+data EmittedGraph
+  = Cyclic [Text]
+  | Acyclic [Text]
+
+debugGraph :: Has (Lift IO) sig m => GraphDebug a => Graphing a -> m ()
+debugGraph gr = sendIO $ do
+  let graphTxt = renderGvFile gr
+  TIO.writeFile gvFileName graphTxt
+{-# WARNING debugGraph "This function is for debug purposes only and should not be used in production." #-}
+
+renderGvFile :: GraphDebug a => Graphing a -> Text
+renderGvFile gr = emit graphEdges
+  where
+    adjMap = unGraphing gr
+    graphEdges = case Acyclic.toAcyclic adjMap of
+      Nothing -> emitGraphEdges adjMap
```

Comment:
Aren't even the cyclic graphs technically directed? I'm thinking it might be useful to know which direction a cycle goes in. Of course, I would probably be looking at the `package-lock.json` file manually anyways so I don't think I really need any changes here - I'm just curious.

Reply:
Good point, we can actually merge these two.  I was thinking of DAG vs. non-DAG, but that's not quite right.

Reply:
I'll merge these quickly, shouldn't take long.

Reply:
Done.